File: commons-mis/src/main/java/io/github/cooperlyt/mis/work/impl/model/WorkModel.java
package io.github.cooperlyt.mis.work.impl.model;

import io.github.cooperlyt.mis.work.data.WorkDefine;
import io.github.cooperlyt.mis.work.data.WorkImpl;
import io.github.cooperlyt.mis.work.message.WorkStatus;
import lombok.*;
import org.apache.commons.lang3.StringUtils;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.annotation.Version;
import org.springframework.data.relational.core.mapping.Table;

import java.time.LocalDateTime;


@EqualsAndHashCode(callSuper = true)
@Table("work")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class WorkModel extends WorkImpl {

  @Builder
  public WorkModel(WorkDefine define, long workId, WorkStatus status, String dataSource) {
    this.workId = workId;
    this.workName = define.getWorkName();
    this.defineId = define.getDefineId();
    this.process = define.isProcess();
    this.type = define.getType();
    this.status = status;
    if (StringUtils.isBlank(dataSource)) {
      this.dataSource = SOURCE_FROM_SYSTEM;
    }else {
      this.dataSource = dataSource;
    }
    if (status.valid){
      this.validateAt = LocalDateTime.now();
    }
    if (WorkStatus.COMPLETED.equals(status)){
      this.completedAt = LocalDateTime.now();
    }
  }

  @Id
  public long getWorkId() {
    return workId;
  }

  @Version
  private Long version;

  @CreatedDate
  public LocalDateTime getCreatedAt(){
    return super.getCreatedAt();
  }

  @LastModifiedDate
  public LocalDateTime getUpdatedAt(){
    return super.getUpdatedAt();
  }

  public WorkModel updateStatus(WorkStatus status){
    if(WorkStatus.COMPLETED.equals(status)){
      super.setCompletedAt(LocalDateTime.now());
      if (this.getValidateAt() == null){
        this.setValidateAt(this.getCompletedAt());
      }
    }
    if (WorkStatus.VALID.equals(status)){
      this.setValidateAt(LocalDateTime.now());
    }
    this.setStatus(status);
    return this;
  }


}

File: commons-mis/src/main/java/io/github/cooperlyt/mis/work/impl/model/WorkModel.kt
package io.github.cooperlyt.mis.work.impl.model

import io.github.cooperlyt.mis.work.data.WorkDefine
import io.github.cooperlyt.mis.work.data.WorkInfo
import io.github.cooperlyt.mis.work.message.WorkStatus
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.Id
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.annotation.Version
import org.springframework.data.relational.core.mapping.Table
import java.time.LocalDateTime

@Table("work")
data class WorkModel(
  @Id
  override val workId: Long,
  override val dataSource: String,
  override val workName: String,
  override var status: WorkStatus,




  override val type: String?,
  override val process: Boolean,
  override val defineId: String,

  override val historyAt: LocalDateTime? = null,
  override var completedAt: LocalDateTime? = null,

  override var validatedAt: LocalDateTime? = null,
  @CreatedDate
  override var createdAt: LocalDateTime? = null,
  @LastModifiedDate
  override var updatedAt: LocalDateTime? = null,
  @Version
  var version: Long? = null
): WorkInfo {

  constructor(define: WorkDefine, workId: Long, status: WorkStatus, dataSource: String, historyAt: LocalDateTime? = null) :this(
    workId = workId,
    dataSource = dataSource,
    workName = define.workName,
    status = status,
    historyAt = historyAt,
    type = define.type,
    process = define.isProcess,
    defineId = define.defineId,
  )
}
